common/src/main/java/dev/isxander/yacl3/gui/controllers/dropdown/AbstractDropdownControllerElement.java

```diff
@@ -0,0 +1,213 @@
+		if (!inputFieldFocused)
+			return false;
+		if (dropdownVisible) {
+			switch (keyCode) {
+				case InputConstants.KEY_DOWN -> {
+					selectNextEntry();
+					return true;
+				}
+				case InputConstants.KEY_UP -> {
+					selectPreviousEntry();
+					return true;
+				}
+				case InputConstants.KEY_TAB -> {
+					if (Screen.hasShiftDown()) {
+						selectPreviousEntry();
+					} else {
+						selectNextEntry();
+					}
+					return true;
+				}
+			}
+		} else {
+			if (keyCode == InputConstants.KEY_RETURN || keyCode == InputConstants.KEY_NUMPADENTER) {
+				showDropdown();
+				return true;
+			}
+		}
+		return super.keyPressed(keyCode, scanCode, modifiers);
+	}
+
+	@Override
+	public boolean charTyped(char chr, int modifiers) {
+		if (!dropdownVisible) {
+			showDropdown();
+		}
+		return super.charTyped(chr, modifiers);
+	}
+
+	@Override
+	protected int getValueColor() {
+		if (inputFieldFocused) {
+			if (!dropdownController.isValueValid(inputField)) {
+				return 0xFFF06080;
+			}
+		}
+		return super.getValueColor();
+	}
+
+	public void selectNextEntry() {
+		if (selectedIndex == getDropdownLength() - 1) {
+			selectedIndex = 0;
+		} else {
+			selectedIndex++;
+		}
+	}
+
+	public void selectPreviousEntry() {
+		if (selectedIndex == 0) {
+			selectedIndex = getDropdownLength() - 1;
+		} else {
+			selectedIndex--;
+		}
+	}
+
+	public int getDropdownLength() {
+		return getMatchingValues().size();
+	}
+
+	public abstract List<U> getMatchingValues();
+
+	public boolean matchingValue(String value) {
+		return value.toLowerCase().contains(inputField.toLowerCase());
+	}
+
+	@Override
+	public void render(GuiGraphics graphics, int mouseX, int mouseY, float delta) {
+		super.render(graphics, mouseX, mouseY, delta);
+
+		if (inputFieldFocused && dropdownVisible) {
+			PoseStack matrices = graphics.pose();
+			matrices.pushPose();
+			matrices.translate(0, 0, 200);
+			renderDropdown(graphics);
+			matrices.popPose();
+		}
+	}
+
+	public void renderDropdown(GuiGraphics graphics) {
+		List<U> options = getMatchingValues();
+		if (options.size() == 0) return;
+		// Limit the visible options to allow scrolling through the suggestion list
+		int begin = Math.max(0, selectedIndex - MAX_SHOWN_NUMBER_OF_ITEMS / 2);
+		int end = begin + MAX_SHOWN_NUMBER_OF_ITEMS;
+		if (end >= options.size()) {
+			end = options.size();
+			begin = Math.max(0, end - MAX_SHOWN_NUMBER_OF_ITEMS);
+		}
+
+		renderDropdownBackground(graphics, end - begin);
+		if (options.size() >= 1) {
+			// Highlight the currently selected element
+			graphics.renderOutline(
+					getDimension().x(),
+					getDimension().yLimit() + 2 + getDimension().height() * (selectedIndex - begin),
+					getDimension().width(),
+					getDimension().height(),
+					-1);
+		}
+
+		int n = 1;
+		for (int i = begin; i < end; ++i) {
+			renderOption(graphics, options.get(i), n);
+			++n;
+		}
+	}
+
+	protected void renderOption(GuiGraphics graphics, U value, int n) {
```

Comment:
Ambiguous with actual `Option`s. Maybe call this something like `renderDropdownEntry`?